tooling/nargo_fmt/src/chunks.rs

```diff
@@ -0,0 +1,973 @@
+    /// We track all this information to see, if we end up needing to format this call
+    /// in multiple lines, if we can write everything up to the left parentheses (inclusive)
+    /// in one line, and just the call arguments in multiple lines.
+    MethodCall {
+        /// This is the width of the group until the left parenthesis (inclusive).
+        width_until_left_paren_inclusive: usize,
+        /// Are there newlines before the left parentheses in this group?
+        has_newlines_before_left_paren: bool,
+        /// Is this method call the left-hand side of a call chain? If so, this is true,
+        /// otherwise this is false an it means it's the outermost call.
+        lhs: bool,
+    },
+    /// The value of an assignment or let statement. We know this is the last group in a chunk so
+    /// if it doesn't fit in the current line but it fits in the next line, we can
+    /// write a newline, indent, and put it there (instead of writing the value in
+    /// multiple lines).
+    AssignValue,
+}
+
+impl GroupKind {
+    fn is_method_call(&self) -> bool {
+        matches!(self, GroupKind::MethodCall { .. })
+    }
+
+    fn is_expression_list(&self) -> bool {
+        matches!(self, GroupKind::ExpressionList { .. })
+    }
+}
+
+/// Interface for creating TextChunks.
+pub(crate) struct ChunkFormatter<'a, 'b>(&'b mut Formatter<'a>);
+
+impl<'a, 'b> ChunkFormatter<'a, 'b> {
+    pub(crate) fn new(formatter: &'b mut Formatter<'a>) -> Self {
+        Self(formatter)
+    }
+
+    /// Stops writing to the current buffer for the duration of the `f` call, which takes
+    /// a formatter to write to. Then, returns a `TextChunk` with the written text.
+    ///
+    /// This allows a caller to format pieces of code and then pre-process them before
+    /// writing to the main buffer.
+    pub(crate) fn chunk(&mut self, f: impl FnOnce(&mut Formatter)) -> TextChunk {
+        let previous_buffer = std::mem::take(&mut self.0.buffer);
+        let previous_indentation = self.0.indentation;
+        self.0.indentation = 0;
+
+        f(self.0);
+
+        self.0.indentation = previous_indentation;
+
+        let buffer = std::mem::replace(&mut self.0.buffer, previous_buffer);
+        TextChunk::new(buffer.contents())
+    }
+
+    /// Stops writing to the current buffer, skips comments and whitespaces (formatting them)
+    /// and returns the formatted result as a `TextChunk`.
+    pub(crate) fn skip_comments_and_whitespace_chunk(&mut self) -> TextChunk {
+        self.chunk(|formatter| {
+            formatter.skip_comments_and_whitespace();
+        })
+    }
+
+    pub(super) fn new_group_tag(&mut self) -> GroupTag {
+        self.0.new_group_tag()
+    }
+
+    pub(super) fn bump(&mut self) -> Token {
+        self.0.bump()
+    }
+}
+
+/// Treating a `ChunkFormatter` as a `Formatter` in read-only mode is always fine,
+/// and reduces some boilerplate.
+impl<'a, 'b> Deref for ChunkFormatter<'a, 'b> {
+    type Target = Formatter<'b>;
```

Comment:
I'm surprised this isn't `'a`, since `ChunkFormatter<'a, 'b>` has a `'b` reference to a `Formatter<'a>` 🤔

Reply:
Oh, good point, it should be `'a`. I don't know why it works, maybe the lifetimes are the same? I'll change it in a follow-up PR after I finish what I'm doing (or feel free to send a PR).